stac_geoparquet/arrow/_to_parquet.py

```diff
@@ -30,14 +32,26 @@ def parse_stac_ndjson_to_parquet(
             infer a common schema across all data and another to read the data and
             iteratively convert to GeoParquet.
     """
-    batches_iter = parse_stac_ndjson_to_arrow(
-        input_path, chunk_size=chunk_size, schema=schema
+    batches = parse_stac_items_to_batches(
+        read_json(input_path), chunk_size=chunk_size, schema=schema
     )
-    first_batch = next(batches_iter)
-    schema = first_batch.schema.with_metadata(_create_geoparquet_metadata())
+    if schema is None:
+        unified_batches = []
+        for batch in batches:
+            if schema is None:
+                schema = batch.schema
+            else:
+                schema = pa.unify_schemas(
+                    [schema, batch.schema], promote_options="permissive"
+                )
+            unified_batches.append(update_batch_schema(batch, schema))
+        batches = unified_batches
+
+    assert schema is not None
+    schema = schema.with_metadata(_create_geoparquet_metadata())
+
     with pq.ParquetWriter(output_path, schema, **kwargs) as writer:
-        writer.write_batch(first_batch)
-        for batch in batches_iter:
+        for batch in batches:
```

Comment:
Just for the record this doesn't work. You can't change the schema you pass to `ParquetWriter` without also changing the physical schema of the arrow data (I don't know how you do that, maybe with `pyarrow.compute.cast` or `table.cast`)

```
import pyarrow.parquet as pq
import pyarrow as pa

table = pa.table({"a": [1, 2, 3, 4]})
assert pa.types.is_int64(table.schema.field("a"))

with pq.ParquetWriter("test.parquet", pa.schema([pa.field("a", pa.int32())])) as writer:
    writer.write_table(table)
```
gives
```

ValueError: Table schema does not match schema used to create file: 
table:
a: int64 vs. 
file:
a: int32
```

Separately, this code was already exhausting the iterator, so it wouldn't have had batches to actually write to parquet